### DISCUSSION TITLE
Screen backlight control by script

### DISCUSSION THREAD

**Discussion by vjurka** [KlipperScreen/KlipperScreen#1164] [Ideas]
This functionality is needed. When the screen saver fires, I need to run a script that turns off the screen backlight, and when the screen comes on, run another script that turns the screen backlight back on. Now, after adjusting the show_screensaver and close_screensavers functions in the screen.py file and adding the line-launching scripts /opt/bl_off.sh and /opt/bl_on.sh, everything works. But when I update KlipperScreen to a newer version those mods are lost.

**Comment by alfrix** [comment]
can you show the content of said scripts? to investigate if this is something that can be added?

**Reply by vjurka** [reply]
First printer scripts:

/etc/scripts/bl_on.sh

_#!/usr/bin/python
import RPi.GPIO as GPIO
GPIO.setmode(GPIO.BCM)
GPIO.setwarnings(False)
GPIO.setup(16,GPIO.OUT)
GPIO.output(16,GPIO.HIGH)_

/etc/scripts/bl_off.sh

_#!/usr/bin/python
import RPi.GPIO as GPIO
GPIO.setmode(GPIO.BCM)
GPIO.setwarnings(False)
GPIO.setup(16,GPIO.OUT)
GPIO.output(16,GPIO.LOW)_

Second printer scripts:

/etc/scripts/bl_on.sh

_/bin/backlight -b 10_


/etc/scripts/bl_off.sh

_/bin/backlight -b 0_

My screen.py changes:

https://github.com/KlipperScreen/KlipperScreen/compare/master...vjurka:KlipperScreen:master?diff=unified&w=#files_bucket

**Comment by alfrix** [comment]
in the meantime you can achieve this by doing this:
~~make a moonraker power device that uses a macro (klipper_device)~~
~~you need to create a macro that runs the scripts (you need gcode_shell_command installable via KIAUH)~~
~~there is an example macro in https://moonraker.readthedocs.io/en/latest/configuration/#power~~
make a moonraker power device that uses [gpio](https://moonraker.readthedocs.io/en/latest/configuration/#gpio-device-configuration)

in ks attach said device to
```
[main]
# Define one or more moonraker power devices that turn on/off with the screensaver (CSV list)
# screen_on_devices: example1
# screen_off_devices:  example1
```

**Reply by vjurka** [reply]
This option is not very suitable. gcode_macro does not work when there is no connection to the MCU. I can't turn on the MCU power supply because the screen is dark and nothing can be seen. If the MCU is powered on, the screen blanking works and the backlight turns on when the screen is touched.

**Reply by vjurka** [reply]
**klipperscreen.log**

_2024-02-15 08:57:46,979 [screen.py:show_panel()] - Reinitializing panel
2024-02-15 08:57:47,119 [screen.py:attach_panel()] - Current panel hierarchy: splash_screen
2024-02-15 08:59:36,439 [screen.py:show_screensaver()] - Showing Screensaver
2024-02-15 08:59:36,448 [screen.py:search_power_devices()] - Found ['EKRANO_ŠVIESA']
2024-02-15 08:59:36,448 [KlippyWebsocket.py:power_device_off()] - Sending machine.device_power.off: EKRANO_ŠVIESA_


**moonraker.log**

_2024-02-15 08:59:36,470 [power.py:set_power()] - Error Toggling Device Power: EKRANO_ŠVIESA
Traceback (most recent call last):
  File "/home/pi/moonraker/moonraker/components/power.py", line 679, in set_power
    await kapis.run_gcode(f"{self.gc_cmd} VALUE={value}")
  File "/home/pi/moonraker/moonraker/components/klippy_apis.py", line 120, in run_gcode
    result = await self._send_klippy_request(
  File "/home/pi/moonraker/moonraker/components/klippy_apis.py", line 108, in _send_klippy_request
    result = await self.klippy.request(req)
  File "/home/pi/moonraker/moonraker/components/klippy_connection.py", line 584, in request
    return await self._request_standard(web_request)
  File "/home/pi/moonraker/moonraker/components/klippy_connection.py", line 678, in _request_standard
    return await base_request.wait(timeout)
  File "/home/pi/moonraker/moonraker/components/klippy_connection.py", line 789, in wait
    return await asyncio.wait_for(asyncio.shield(self._fut), to)
  File "/usr/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
moonraker.utils.ServerError: mcu 'mcu': Unable to connect
Once the underlying issue is corrected, use the
"FIRMWARE_RESTART" command to reset the firmware, reload the
config, and restart the host software.
Error configuring printer

2024-02-15 08:59:36,470 [common.py:build_error()] - JSON-RPC Request Error - Requested Method: machine.device_power.off, Code: 400, Message: Error Toggling Device Power: EKRANO_ŠVIESA_

**Reply by alfrix** [reply]
you can configure a moonraker power device that uses a gpio directly without a macro

https://moonraker.readthedocs.io/en/latest/configuration/#gpio-device-configuration

**Reply by vjurka** [reply]
On a printer where the screen backlight is controlled via gpio, using screen_on_devices: screen_off_devices: works correctly. In another printer, the backlight of the screen is controlled via the screen manufacturer's app via a USB connection, control via a script is required there, a macro is needed to run the script, and the macro does not work if there is no connection to the MCU.
